src/full/Agda/TypeChecking/Sort.hs

```diff
@@ -86,27 +87,24 @@ hasBiggerSort = void . inferUnivSort
 {-# SPECIALIZE inferPiSort :: Dom Type -> Abs Sort -> TCM Sort #-}
 -- | Infer the sort of a Pi type.
 --   If we can compute the sort straight away, return that.
---   Otherwise, return a 'PiSort' and add a constraint to ensure we can compute the sort eventually.
---
+--   Otherwise, return a 'PiSort'.
+--   Note that this function does NOT check PTS constraints, use hasPTSRule for that
 inferPiSort :: (PureTCM m, MonadConstraint m)
   => Dom Type  -- ^ Domain of the Pi type.
   -> Abs Sort  -- ^ (Dependent) sort of the codomain of the Pi type.
   -> m Sort    -- ^ Sort of the Pi type.
-inferPiSort a s = do
-  s1' <- reduceB $ getSort a
-  s2' <- mapAbstraction a reduceB s
-  let s1 = ignoreBlocking s1'
-  let s2 = ignoreBlocking <$> s2'
-  --Jesper, 2018-04-23: disabled PTS constraints for now,
-  --this assumes that piSort can only be blocked by unsolved metas.
-  --Arthur Adjedj, 2023-02-27, Turned PTS back on,
-  --piSort can now be blocked by Leveluniv
-  case piSort' (unEl <$> a) s1 s2 of
-    Right s -> return s
-    Left b -> do
-      let b' = unblockOnEither (getBlocker s1') (getBlocker $ unAbs s2')
-      addConstraint (unblockOnEither b b') $ HasPTSRule a s2
-      return $ PiSort (unEl <$> a) s1 s2
+inferPiSort a s2 = case piSort' at s1 s2 of
+  Left _ -> return $ PiSort at s1 s2
+  -- Jesper, 2025-09-15: if a PiSort reduces to a FunSort, piSort'
+  -- just returns the FunSort without trying to simplify it further.
+  -- So if we get a FunSort here we call inferFunSort in the hopes
+  -- of getting a simpler result. But we DON'T call reduce as that
+  -- can lead to quadratic behavior, see #8096.
+  Right (FunSort s1' s2') -> inferFunSort (El s1' <$> at) s2'
+  Right s' -> return s'
+  where
+    at = unEl <$> a
```

Comment:
`at` reads a bit funny and it a new naming style, how about `va` (existing naming style)?